### DISCUSSION TITLE
a-failed-experiment-with-interceptors-in-csharp-12-and-dotnet-8/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#397] [Announcements]
# a-failed-experiment-with-interceptors-in-csharp-12-and-dotnet-8/

In this post I described how I tried (and failed) to add an interceptor for Enum.ToString(), due to a bug in the Roslyn compiler.

https://andrewlock.net/a-failed-experiment-with-interceptors-in-csharp-12-and-dotnet-8/

**Comment by Xor-el** [comment]
Hello Andrew, thanks for this very insightful write-up, learnt quite a lot from it.
meanwhile, minor typo here

of `Enum.IsDefined<Colour(value)`. That's where interceptors come in.

should be

of `Enum.IsDefined<Colour>(value)`. That's where interceptors come in.

**Comment by cocowalla** [comment]
Another good option for faster enum operations is Enums.NET (https://github.com/TylerBrinkley/Enums.NET).

**Comment by felpel** [comment]
In the meantime, could you develop your own Roslyn analyzers so that you can detect all aforementioned scenarios and warn/inform users that they should use the faster version?

**Reply by andrewlock** [reply]
Yep, that's definitely an option. Would definitely take a PR for that if someone wanted to do it 😇

**Comment by majellin24** [comment]
Great article as always, just one point of clarification.

>Interceptors are designed to be used with source generators, in that they require you to specify the exact file, line, and character number of the method you're going to replace.

I'm pretty sure you mean that interceptors are designed to be *generated by* source generators?  In that the source generator would be able to pick out the line and column during analysis and then generate each corresponding attribute on the source generated method.

That was my interpretation anyway, just wanted to see if that was correct as I've run into trouble making assumptions about roslyn in the past, thanks!

**Reply by andrewlock** [reply]
Yep exactly. You will typically source-generate the interceptor 👍

**Comment by bkoelman** [comment]
I'm curious if your interceptor is also invoked in cases such as:
```c#
Console.WriteLine($"You chose {colour}");
```
Have you tried that?
